src/number_theory/prime_counting.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2021 Bolton Bailey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey
+-/
+
+import data.nat.prime
+import data.nat.totient
+import algebra.periodic
+import data.finset.locally_finite
+import data.nat.count
+
+
+/-!
+# The Prime Counting Function
+
+In this file we define the prime counting function - the function on natural numbers that returns
+the number of primes less than or equal to its input.
+
+## Main Results
+
+The main definitions for this file are
+
+- `prime_counting`: The prime counting function π
+- `prime_counting'`: π(n - 1)
+
+We then prove that these are monotone. The last main theorem is an upper bound on `π'` which arises
+by observing that all numbers greater than `k` and not coprime to `k` are not prime, and so only at
+most `φ(k)/k` fraction of the numbers from `k` to `n` are prime.
+
+## Notation
+
+We use the standard notation `π` to represent the prime counting function (and `π'` to represent
+the reindexed version).
+
+-/
+
+namespace nat
+open finset
+
+/--
+A variant of the traditional prime counting function which gives the number of primes
+*strictly* less than the input. More convenient for avoiding off-by-one errors.
+-/
+def prime_counting' : ℕ → ℕ := nat.count prime
+
+/-- The prime counting function: Returns the number of primes less than or equal to the input. -/
+def prime_counting (n : ℕ) : ℕ := prime_counting' (n + 1)
+
+localized "notation `π` := nat.prime_counting" in nat
+localized "notation `π'` := nat.prime_counting'" in nat
+
+lemma monotone_prime_counting' : monotone prime_counting' := count_monotone prime
+
+lemma monotone_prime_counting : monotone prime_counting :=
+λ a b a_le_b, monotone_prime_counting' (add_le_add_right a_le_b 1)
+
+private lemma filter_coprime_bound (a k n : ℕ) (a_pos : 0 < a) :
```

Comment:
```suggestion
private lemma Ico_filter_coprime_le (a k n : ℕ) (a_pos : 0 < a) :
```

Reply:
I don't think this should be a private lemma. Please make it a public lemma and move it to the file on `totient`.